src/Features/Core/Portable/Diagnostics/Service/DiagnosticAnalyzerService.cs

```diff
@@ -122,21 +124,102 @@ public async Task<ImmutableArray<DiagnosticData>> GetDiagnosticsForSpanAsync(
     {
-        return _incrementalAnalyzer.GetDiagnosticsForIdsAsync(project, documentId, diagnosticIds, shouldIncludeAnalyzer, includeLocalDocumentDiagnostics, includeNonLocalDocumentDiagnostics, cancellationToken);
+        var analyzers = await GetDiagnosticAnalyzersAsync(
+            project, diagnosticIds, shouldIncludeAnalyzer, cancellationToken).ConfigureAwait(false);
+
+        return await ProduceProjectDiagnosticsAsync(
+            project, analyzers, diagnosticIds,
+            // Ensure we compute and return diagnostics for both the normal docs and the additional docs in this
+            // project if no specific document id was requested.
+            documentId != null ? [documentId] : [.. project.DocumentIds, .. project.AdditionalDocumentIds],
+            includeLocalDocumentDiagnostics,
+            includeNonLocalDocumentDiagnostics,
+            // return diagnostics specific to one project or document
+            includeProjectNonLocalResult: documentId == null,
+            cancellationToken).ConfigureAwait(false);
     }
 
-    public Task<ImmutableArray<DiagnosticData>> GetProjectDiagnosticsForIdsAsync(
+    public async Task<ImmutableArray<DiagnosticData>> GetProjectDiagnosticsForIdsAsync(
         Project project, ImmutableHashSet<string>? diagnosticIds,
         Func<DiagnosticAnalyzer, bool>? shouldIncludeAnalyzer,
         bool includeNonLocalDocumentDiagnostics, CancellationToken cancellationToken)
     {
-        return _incrementalAnalyzer.GetProjectDiagnosticsForIdsAsync(project, diagnosticIds, shouldIncludeAnalyzer, includeNonLocalDocumentDiagnostics, cancellationToken);
+        var analyzers = await GetDiagnosticAnalyzersAsync(
+            project, diagnosticIds, shouldIncludeAnalyzer, cancellationToken).ConfigureAwait(false);
+
+        return await ProduceProjectDiagnosticsAsync(
+            project, analyzers, diagnosticIds,
+            documentIds: [],
+            includeLocalDocumentDiagnostics: false,
+            includeNonLocalDocumentDiagnostics: includeNonLocalDocumentDiagnostics,
+            includeProjectNonLocalResult: true,
+            cancellationToken).ConfigureAwait(false);
+    }
+
+    private Task<ImmutableArray<DiagnosticData>> ProduceProjectDiagnosticsAsync(
+        Project project,
+        ImmutableArray<DiagnosticAnalyzer> analyzers,
+        ImmutableHashSet<string>? diagnosticIds,
+        IReadOnlyList<DocumentId> documentIds,
+        bool includeLocalDocumentDiagnostics,
+        bool includeNonLocalDocumentDiagnostics,
+        bool includeProjectNonLocalResult,
+        CancellationToken cancellationToken)
+    {
+        return _incrementalAnalyzer.ProduceProjectDiagnosticsAsync(
+            project, analyzers, diagnosticIds, documentIds,
+            includeLocalDocumentDiagnostics,
+            includeNonLocalDocumentDiagnostics,
+            includeProjectNonLocalResult,
+            cancellationToken);
+    }
```

Comment:
this method will now be the one that ccan call over to oop immeidately.  we had to do some in-memory processing before calling it (specifically because lambdas are passed in), so we need to still do that on the host side.  but none of that involves compilations, and we can now remote this call entirely over, and have the remote side call into this.